### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by osmanwako**
# Inside this PR We did a system test and fix N+! problems
- We solved the N+1 problem when fetching all posts and their comments for a use
  - when the user fetches **user** index page before fixing N+! problems
userb1
  - when the user fetches **user** index page after fixing N+! problems
usera1
  -  when the user fetches **user** show page before fixing N+! problems
userb1
 - when the user fetches **user** show page after fixing N+! problems
userb2
  - when the user fetches **post** index page before fixing N+! problems
postb1
  - when the user fetches **post** index page after fixing N+! problems
posta1
  -  when the user fetches **post** show page before fixing N+! problems
postb2
  - when the user fetches **post** show page after fixing N+! problems
posta2

### System Test
- Two test files were created inside **test/system** (users_test and posts_test) directory
 - To run ** rails test:system**

### CODE REVIEW

File: test/system/users_test.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    assert_selector 'a', text: @userone.name
+    assert_selector 'p', text: "Number of post #{@userone.posts_counter}"
+    assert 'avator.png', @userone.photo
+
+    assert_selector 'a', text: @usertwo.name
+    assert_selector 'p', text: "Number of post #{@usertwo.posts_counter}"
+    assert 'avator.png', @usertwo.photo
+    click_link @userone.name, match: :first
+    assert user_path(@userone)
+  end
```
Comment (Lizdev-05) [inline test/system/users_test.rb:19]: - [ ] Please make sure that you have a Spec file that only test `User` index page :pray:

File: test/system/users_test.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    end
+    assert_selector 'a', text: 'See all Posts'
+    click_link 'See all Posts'
+    assert user_posts_path(@usertwo)
+    visit user_path(@usertwo)
+    post = @usertwo.recent_posts[0]
+    assert_selector 'a', text: post.title
+    click_link post.title, match: :first
+    assert user_post_path(@usertwo, post)
+  end
```
Comment (Lizdev-05) [inline test/system/users_test.rb:39]: - [ ] Please make sure that you have a Spec file that only tests the `User` show page :pray:

**Review by zieeco — APPROVED** [review]
# status: APPROVED :green_circle: :100: 

Hi, @Team, 






Good job so far 👍 Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 


## Highlights ⚡ 

✅ Descriptive PR 👍 
✅ Fix n+1 problem 🔥 
✅ Linters are passing 👏 


## Optional suggestions
- N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._



Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: test/system/posts_test.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      assert_selector 'a', text: post.title
+      assert_selector 'p', text: post.text
+      assert_selector 'span', text: post.comments.first.text
+      assert_selector 'span', text: post.comments_counter
+      assert_selector 'span', text: post.likes_counter
+    end
+    assert_selector 'a', text: 'Pagination'
+    click_link @userone.posts.first.title, match: :first
+    assert user_post_path(@userone, @userone.posts.first)
+  end
```
Comment (Lizdev-05) [inline test/system/posts_test.rb:26]: - [ ] Please make sure that you have a Spec file that only tests the `Post` index page :pray:

File: test/system/posts_test.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    assert_selector 'strong', text: @usertwo.name
+    assert_selector 'strong', text: @usertwo.posts.first.title
+    assert_selector 'span', text: @usertwo.posts.first.comments_counter
+    assert_selector 'span', text: @usertwo.posts.first.likes_counter
+    assert_selector 'p', text: @usertwo.posts.first.text
+    @usertwo.posts.first.comments.each do |comment|
+      assert_selector 'span', text: comment.user.name
+      assert_selector 'span', text: comment.text
+    end
+  end
```
Comment (Lizdev-05) [inline test/system/posts_test.rb:39]: - [ ] Please make sure that you have a Spec file that only tests the `Post` show page :pray:

**Review by Lizdev-05 — CHANGES_REQUESTED** [review]
Hi @osmanwako     and @vanina123  , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
:heavy_check_mark: Professional Readme file :+1: 
:heavy_check_mark: Good commit history :+1:
:heavy_check_mark: No linter error :+1: 

## Required Changes ♻️
_Check the comments under the review._

- [ ] Please in the project you are supposed to write integration tests with [Capybara gem](https://github.com/teamcapybara/capybara) to cover the user workflows as you test all of their views so I think you have Capybara gem in `Gemfile` file for integration test :blush: 
  
  ```ruby
   group :development, :test do
     ...
     **gem 'capybara'**
     ...
   end
  ```

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Lizdev-05** [comment]
- [ ] Hi @osmanwako and @vanina123 , When I was reviewing and submitting the suggested changes review on this pull request, your project was not approved and not merged but after I submit the review, I noticed it is approved and merged so what is going on?

**Comment by osmanwako** [comment]
Hi @Lizdev-05 Elizabeth
How did you get the project accepted by an other CR first?

File: test/system/posts_test.rb (hunk 4 above)
Comment (osmanwako) [reply test/system/posts_test.rb:39]: Hi @Lizdev-05  Elizabeth
we did 40 test cases in this project.
inside the user index page we wrote test case assure
    - User index page: visit user_path assert
  - we can see the username of all other users.
  - we can see the profile picture for each user.
  - we can see the number of posts each user has written.
  - When we click on a user, we am redirected to that user's show page.
-  Inside User  show index 
     - we can see the user's profile picture.
  - we can see the user's username.
  - we can see the number of posts the user has written.
  - we can see the user's bio.
  - we can see the user's first 3 posts.
  - we can see a button that lets me view all of a user's posts.
  - When we click a user's post, it redirects me to that post's show page.
  - When we  click to see all posts, it redirects me to the user's post's index page.

Comment (osmanwako) [reply test/system/posts_test.rb:39]: We follow the resource given by microverse [my resource ](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/integrations_lesson.md)
- Check out [this tutorial](https://www.youtube.com/watch?v=2TInLtG8dj4) for using Capybara with RSpec in Rails 7.

**Pull request merged**